client/mifarehost.c

```diff
@@ -840,16 +883,62 @@ int mfTraceDecode(uint8_t *data_src, int len, bool wantSaveToEmlFile) {
 				lfsr_rollback_word(revstate, uid ^ nt, 0);
 
 				crypto1_get_lfsr(revstate, &lfsr);
-				printf("key> %x%x\n", (unsigned int)((lfsr & 0xFFFFFFFF00000000) >> 32), (unsigned int)(lfsr & 0xFFFFFFFF));
+				ui64Key = lfsr;
+				printf("key> %x%x Prng:%s\n", 
+					(unsigned int)((lfsr & 0xFFFFFFFF00000000) >> 32), (unsigned int)(lfsr & 0xFFFFFFFF), 
+					validate_prng_nonce(nt) ? "WEAK": "HARDEND");
 				AddLogUint64(logHexFileName, "key> ", lfsr);
 			} else {
-				printf("key> nested not implemented!\n");
-				at_enc = bytes_to_num(data, 4);
+				if (validate_prng_nonce(nt)) {
+					struct Crypto1State *pcs;
+					pcs = crypto1_create(ui64Key);
+					uint32_t nt1 = crypto1_word(pcs, nt_enc ^ uid, 1) ^ nt_enc;
+					uint32_t ar = prng_successor(nt1, 64);
+					uint32_t at = prng_successor(nt1, 96);
+					printf("key> nested auth uid: %08x nt: %08x nt_parity: %s ar: %08x at: %08x\n", uid, nt1, printBitsPar(&nt_enc_par, 4), ar, at);
+					uint32_t nr1 = crypto1_word(pcs, nr_enc, 1) ^ nr_enc;
+					uint32_t ar1 = crypto1_word(pcs, 0, 0) ^ ar_enc;
+					uint32_t at1 = crypto1_word(pcs, 0, 0) ^ at_enc;
+					printf("key> the same key test. nr1: %08x ar1: %08x at1: %08x \n", nr1, ar1, at1);
+
+					if (NTParityCheck(nt1))
+						printf("key> the same key test OK. key=%x%x\n", (unsigned int)((ui64Key & 0xFFFFFFFF00000000) >> 32), (unsigned int)(ui64Key & 0xFFFFFFFF));
+					else
+						printf("key> the same key test. check nt parity error.\n");
+					
+					uint32_t ntx = prng_successor(nt, 90);

```

Comment:
`nt`? Where does it come from? Isn't this the nt from the first authentication?  You  must not assume a fixed time difference between first and second authentication. This is only  true withing `hf mf nested` but not when sniffing a communication between reader and card.

Reply:
it comes from previous authentication. some cards not cycle prng between transmit and if we have 2 authentications in a row there will be not so much distance.
full prng cycle needs aprox 600ms.

there is just a speed optimization.

Reply:
How do you know how fast the nested authentication follows the first authentication? There can be 0 or many block reads in between them. And you don't know how fast the reader issues commands.

Reply:
How do you know how fast the nested authentication follows the first authentication? There can be 0 or many block reads in between them. And you don't know how fast the reader issues commands.

Reply:
I dont know, but usually it faster than 600ms. Its just speed optimization. we can check it from first, but for speed we start from last one

Reply:
OK. Agreed.